### DISCUSSION TITLE
Install a specific PHP patch/minor version?

### DISCUSSION THREAD

**Discussion by jaydrogers** [oerdnj/deb.sury.org#2014] [Q&A]
Hello,

I searched Google, asked on Twitter, searched the issue/discussions, and I could not find a clear answer.

### 👉 Is it possible to install a PHP version down to the patch/minor number?

**For example:**
- PHP Version:`8.2.9` vs `8.2.8`

I read [Installing and maintaining a specific version release](https://github.com/oerdnj/deb.sury.org/wiki/Frequently-Asked-Questions#installing-and-maintaining-a-specific-version-release) and [Managing Multiple Versions](https://github.com/oerdnj/deb.sury.org/wiki/Managing-Multiple-Versions) but I could not find a clear answer.

### My assumption
- I believe the answer is "no"

### Use case
I maintain open source PHP Docker images that use this PPA to install PHP. A handful of users have requested pinning to the minor version for better stability.

**Comment by bashgeek** [comment]
Theoretically you could do something like this with APT pinning the correct version you want, but this won't work here. With every new version the older versions packages get deleted, so you cannot download them anymore.

So the only thing you could do is, when you've already have the version you want installed, put an APT hold on it to prevent it from getting updated.

**Reply by jaydrogers** [reply]
Interesting! Thanks for the suggestion.

With pinning the version with APT, do you have an example or more resources where I could learn more about this?

I could possibly do this:
1. Get an official PHP version (with patch number)
1. Pass it to APT
1. Build an image
1. Tag the image with the patch number

The image will "snapshot" that version forever. I would just have to do this while that version is "active".

The user could run `apt update && apt upgrade` if they chose to, but they would also be guaranteed that version as well.

I appreciate your help and insight!

**Reply by spyridonas** [reply]
Maybe this is the thing you are after ?
https://github.com/crazywhalecc/static-php-cli
It's way more complicated, basically it builds the php you want, but the resulting binary is a single file and can be copied with docker build steps

**Reply by jaydrogers** [reply]
Thanks for chiming in! If I understand this correctly, this would basically bypass Ondrej PHP?

**Reply by spyridonas** [reply]
Yes,
It will download the php source and compile it
Then you end up with a single binary that can be copied over.
It's time consuming but you can give a very specific version of php
The docs have a great example, 

bin/spc download --all -U "php-src:https://downloads.php.net/~eric/php-8.3.0beta1.tar.gz"

**Reply by jaydrogers** [reply]
Thank you! I will explore this as an option.

My preference would be to stay with Ondrej PHP if I could just so it would be easier for people to extend.

For some background, I am maintaining these Docker images and just want a user friendly way for people to pin PHP versions and add their own extensions if needed: https://github.com/serversideup/docker-php/

I appreciate your feedback, thanks again! 🙌

**Reply by spyridonas** [reply]
I knew I've seen this name before!
https://github.com/serversideup/docker-php/pull/185
Ha. Anyways, I believe that my solution doesn't work for you, but if you look at the official php docker images this is how they do it (they download and compile the specific php version). 

https://github.com/docker-library/php/blob/master/8.2/alpine3.18/fpm/Dockerfile

**Reply by jaydrogers** [reply]
Thanks for your feedback! Lot's of great things coming up soon. Feel free to chime in as I build more: https://github.com/serversideup/docker-php/pull/205

I'm currently working to refactor a bunch of things to improve stability, extensibility, and compatibility.

I greatly appreciate you sharing your thoughts and I have a lot more things to chew on 😃
